File: Plugins/QuorumPlugins/src/main/java/plugins/quorum/Libraries/Language/Types/Number.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package plugins.quorum.Libraries.Language.Types;

import quorum.Libraries.Language.Object_;
import quorum.Libraries.Language.Types.Number_;

/**
 * Number plugin for handling Number class functions for objects and primitives.
 * 
 * @author Melissa Stefik
 */
public class Number {
    public java.lang.Object me_ = null;
    private double number = -1;
    
    public static double ParseNumber(String value) throws quorum.Libraries.Language.Errors.CastError{
        try {
            return java.lang.Double.parseDouble(value);
        } catch(NumberFormatException e) {
            throw new quorum.Libraries.Language.Errors.CastError();
        }
    }
    
    public static double ConvertObjectToNumber(Object_ object) {
        if(object == null) {
            throw new NullPointerException("Cannot convert an undefined type to an integer.");
        }
        Number_ val = (Number_) object;
        return val.GetValue();
    }
    
    public static Object_ ConvertNumberToObject(double value) {
        Number_ i = new quorum.Libraries.Language.Types.Number();
        i.SetValue(value);
        return i;
    }
    
    public static double ConvertNumberObjectToNumber(Number_ object) {
        if(object == null) {
            throw new NullPointerException("Cannot convert an undefined type to an integer.");
        }
        return object.GetValue();
    }
    
    public static Number_ ConvertNumberToNumberObject(double value) {
        Number_ i = new quorum.Libraries.Language.Types.Number();
        i.SetValue(value);
        return i;
    }
    
    public double GetMaximumValue(){
        return Double.MAX_VALUE;
    }
    
    public static double PrimitiveGetMaximumValue(double self){
        return Double.MAX_VALUE;
    }
    
    public double GetMinimumValue(){
        return -Double.MAX_VALUE;
    }
    
    public static double PrimitiveGetMinimumValue(double self){
        return -Double.MAX_VALUE;
    }
    
    public double GetMinimumPositiveValue(){
        return Double.MIN_VALUE;
    }
    
    public static double PrimitiveGetMinimumPositiveValue(double self){
        return Double.MIN_VALUE;
    }
    
    public double GetNotANumberValue(){
        return Double.NaN;
    }
    
    public static double PrimitiveGetNotANumberValue(double self){
        return Double.NaN;
    }
    
    public double GetNegativeInfinityValue(){
        return Double.NEGATIVE_INFINITY;
    }
    
    public static double PrimitiveGetNegativeInfinityValue(double self){
        return Double.NEGATIVE_INFINITY;
    }
    
    public double GetPositiveInfinityValue(){
        return Double.POSITIVE_INFINITY;
    }
    
    public static double PrimitiveGetPositiveInfinityValue(double self){
        return Double.POSITIVE_INFINITY;
    }
    
    public double GetNumberOfBits(){
        return Double.SIZE;
    }
    
    public static int PrimitiveGetNumberOfBits(double self){
        return Double.SIZE;
    }
    
    public void SetValueNative(double value){
        number = value;
    }
    
    public static boolean PrimitiveEquals(double self, Object_ obj) throws quorum.Libraries.Language.Errors.Error{
        Number_ t = (Number_)obj;
        return self == t.GetValue();
    }
    
    public static int PrimitiveGetInteger(double self){
        return (int)self;
    }
    
    public static String PrimitiveGetText(double self){
        return "" + self;
    }
    
    public static int PrimitiveCompare(double self, quorum.Libraries.Language.Object_ obj){
        Number_ t = (Number_)obj;
        if(self == t.GetValue()){
            return 0;
        } else if(self > t.GetValue()){
            return 1;
        } else {
            return -1;
        }
    }
    
    public String GetHex(){
        return java.lang.Double.toHexString(number);
    }
    
    public static String PrimitiveGetHex(double self){
        return java.lang.Double.toHexString(self);
    }
    
    public boolean IsInfinite(){
        return java.lang.Double.isInfinite(number);
    }
    
    public static boolean PrimitiveIsInfinite(double self){
        return java.lang.Double.isInfinite(self);
    }
    
    public boolean IsNotANumber(){
        return java.lang.Double.isNaN(number);
    }
    
    public static boolean PrimitiveIsNotANumber(double self){
        return java.lang.Double.isNaN(self);
    }
    
    public int GetHashCode() {
        long returnMe = Double.doubleToLongBits(number);
        return (int)(returnMe^(returnMe>>>32));
    }
    
    public int PrimitiveGetHashCode(double self) {
        long returnMe = Double.doubleToLongBits(self);
        return (int)(returnMe^(returnMe>>>32));
    }
}

File: Quorum/Library/Standard/Plugins/javascript/Libraries/Language/Types/Number.js
class plugins_quorum_Libraries_Language_Types_Number_ {
    constructor(optional) {
      this.number = 0.0;

      if(optional === undefined) {
      } else {
          number = optional;
      }
    };
    
    
    SetValueNative$quorum_number(value) {
        number = value;
    };
    //system action GetMaximumValue returns number
    GetMaximumValue() {
        return this.number.GetMaximumValue$NumberPrimitive();
    };
    //system action GetMinimumValue returns number
    GetMinimumValue() {
        return this.number.GetMinimumValue$NumberPrimitive();
    };
    //system action GetMinimumPositiveValue returns number
    GetMinimumPositiveValue() {
        return this.number.GetMinimumPositiveValue$NumberPrimitive();
    };
    //system action GetPositiveInfinityValue returns number
    GetPositiveInfinityValue() {
        return this.number.GetPositiveInfinityValue$NumberPrimitive();
    };
    //system action GetNumberOfBits returns integer
    GetNumberOfBits() {
        return this.number.GetNumberOfBits$NumberPrimitive();
    };
    //system action GetNegativeInfinityValue returns number
    GetNegativeInfinityValue() {
        return this.number.GetNegativeInfinityValue$NumberPrimitive();
    };
    //system action GetNotANumberValue returns number
    GetNotANumberValue() {
        return this.number.GetNotANumberValue$NumberPrimitive();
    };
    //system action IsInfinite() returns boolean
    IsInfinite() {
        return this.number.IsInfinite$NumberPrimitive();
    };
    //system action IsNotANumber() returns boolean
    IsNotANumber() {
        return this.number.IsNotANumber$NumberPrimitive();
    };
    //system action GetHex() returns text
    GetHex() {
        return this.number.GetHex$NumberPrimitive();
    };
    //system action GetHashCode() returns integer
    GetHashCode() {
        return this.number.GetHashCode$NumberPrimitive();
    };
}

//system action GetMaximumValue returns number
Number.prototype.GetMaximumValue$NumberPrimitive = function() {
  return Number.MAX_VALUE;
};
//system action GetMinimumValue returns number
Number.prototype.GetMinimumValue$NumberPrimitive = function() {
  return -Number.MAX_VALUE;
};
//system action GetMinimumPositiveValue returns number
Number.prototype.GetMinimumPositiveValue$NumberPrimitive = function() {
       Number.MIN_VALUE;
};  
//system action GetPositiveInfinityValue returns number
Number.prototype.GetPositiveInfinityValue$NumberPrimitive = function() {
  return Number.POSITIVE_INFINITY;
};
//system action GetNumberOfBits returns integer
Number.prototype.GetNumberOfBits$NumberPrimitive = function() {
  return 0;
};
//system action GetNegativeInfinityValue returns number
Number.prototype.GetNegativeInfinityValue$NumberPrimitive = function() {
  return Number.NEGATIVE_INFINITY;
};
//system action GetNotANumberValue returns number
Number.prototype.GetNotANumberValue$NumberPrimitive = function() {
  return Number.NaN;
};
//system action IsInfinite() returns boolean
Number.prototype.IsInfinite$NumberPrimitive = function() {
    return !isFinite(this.valueOf());
};
//system action IsNotANumber() returns boolean
Number.prototype.IsNotANumber$NumberPrimitive = function() {
  return isNaN(this.valueOf());
};
//system action GetHex() returns text
Number.prototype.GetHex$NumberPrimitive = function() {
  return this.valueOf().toString(16);
};
//system action GetHashCode() returns integer
Number.prototype.GetHashCode$NumberPrimitive = function() {
  return Math.floor(this.valueOf());
};

Number.prototype.GetInteger$NumberPrimitive = function() {
  return Math.floor(this.valueOf());
};
Number.prototype.GetText$NumberPrimitive = function() {
  var illegal_ = this.valueOf();
  var res = (illegal_ % 1 == 0 ? illegal_.toFixed(1) : illegal_);
  return res;
};
Number.prototype.Compare$quorum_Libraries_Language_Object$NumberPrimitive = function(value) {
    var b = global_CheckCast(value, "Libraries.Language.Types.Number");
    
    var me = this.valueOf();
    var other = b.GetValue();
    if ((me == other)) {
        return 0;
    }
    else if ((me < other)) {
        return -1;
    }
    else {
        return 1;
    }
};
Number.prototype.Equals$quorum_Libraries_Language_Object$NumberPrimitive = function (value) {
    var b = global_CheckCast(value, "Libraries.Language.Types.Number");

    var me = this.valueOf();
    var other = b.GetValue();
    if ((me == other)) {
        return true;
    } else {
        return false;
    }
};


